Rollup of 11 pull requests

kennytm opened the pull request:
- Successful merges: #48639, #49223, #49333, #49369, #49381, #49395, #49399, #49401, #49417, #49202, #49426
- Failed merges:

varkor committed:
Compute each key only one during slice::sort_by_key

varkor committed:
Update documentation for sort_by_key

varkor committed:
Clarify behaviour of sort_unstable_by_key with respect to sort_by_key

varkor committed:
Add a test for sort_by_key

varkor committed:
Update documentation

varkor committed:
Cull the quadratic

varkor committed:
Index enumeration by minimally sized type

varkor committed:
Add sort_by_cached_key method

varkor committed:
Fix use of unstable feature in test

varkor committed:
Improve and fix documentation for sort_by_cached_key

varkor committed:
Clarify time complexity

varkor committed:
Check that the size optimisation is not redundant

varkor committed:
Add lexicographic sorting benchmark

varkor committed:
Add stability test for sort_by_cached_key

GuillaumeGomez committed:
Propose a variant if it is an enum for E0599

GuillaumeGomez committed:
Fix impl assoc constant link not working

petrochenkov committed:
Fix pretty-printing for raw identifiers

Commit:
Add is_whitespace and is_alphanumeric to str.

The other methods from `UnicodeStr` are already stable inherent
methods on str, but these have not been included.

Commit:
Remove unnecessary trait import.

varkor committed:
Remove mentions of unstable sort_by_cached key from stable documentation

petrochenkov committed:
libsyntax: Remove obsolete.rs

ehuss committed:
Fix diagnostic colors on Windows 10 console.

This updates termcolor to pick up BurntSushi/ripgrep#867.

Fixes #49322.

Commit:
Add missing '?' to format grammar.

kennytm committed:
Rollup merge of #48639 - varkor:sort_by_key-cached, r=bluss

Add slice::sort_by_cached_key as a memoised sort_by_key

At present, `slice::sort_by_key` calls its key function twice for each comparison that is made. When the key function is expensive (which can often be the case when `sort_by_key` is chosen over `sort_by`), this can lead to very suboptimal behaviour.

To address this, I've introduced a new slice method, `sort_by_cached_key`, which has identical semantic behaviour to `sort_by_key`, except that it guarantees the key function will only be called once per element.

Where there are `n` elements and the key function is `O(m)`:
- `slice::sort_by_cached_key` time complexity is `O(m n log m n)`, compared to `slice::sort_by_key`'s `O(m n + n log n)`.
- `slice::sort_by_cached_key` space complexity remains at `O(n + m)`. (Technically, it now reserves a slice of size `n`, whereas before it reserved a slice of size `n/2`.)

`slice::sort_unstable_by_key` has not been given an analogue, as it is important that unstable sorts are in-place, which is not a property that is guaranteed here. However, this also means that `slice::sort_unstable_by_key` is likely to be slower than `slice::sort_by_cached_key` when the key function does not have negligible complexity. We might want to explore this trade-off further in the future.

Benchmarks (for a vector of 100 `i32`s):
```
# Lexicographic: `|x| x.to_string()`
test bench_sort_by_key ... bench:      112,638 ns/iter (+/- 19,563)
test bench_sort_by_cached_key ... bench:       15,038 ns/iter (+/- 4,814)

# Identity: `|x| *x`
test bench_sort_by_key ... bench:        1,346 ns/iter (+/- 238)
test bench_sort_by_cached_key ... bench:        1,839 ns/iter (+/- 765)

# Power: `|x| x.pow(31)`
test bench_sort_by_key ... bench:        3,624 ns/iter (+/- 738)
test bench_sort_by_cached_key ... bench:        1,997 ns/iter (+/- 311)

# Abs: `|x| x.abs()`
test bench_sort_by_key ... bench:        1,546 ns/iter (+/- 174)
test bench_sort_by_cached_key ... bench:        1,668 ns/iter (+/- 790)
```
(So it seems functions that are single operations do perform slightly worse with this method, but for pretty much any more complex key, you're better off with this optimisation.)

I've definitely found myself using expensive keys in the past and wishing this optimisation was made (e.g. for https://github.com/rust-lang/rust/pull/47415). This feels like both desirable and expected behaviour, at the small cost of slightly more stack allocation and minute degradation in performance for extremely trivial keys.

Resolves #34447.

kennytm committed:
Rollup merge of #49223 - GuillaumeGomez:propose-variant-for-E0599, r=cramertj

Propose a variant if it is an enum for E0599

Fixes #49192.

kennytm committed:
Rollup merge of #49333 - GuillaumeGomez:link-assoc-const, r=QuietMisdreavus

Fix impl assoc constant link not working

Fixes #49323.

r? @QuietMisdreavus

kennytm committed:
Rollup merge of #49369 - petrochenkov:rprint, r=oli-obk

Fix pretty-printing for raw identifiers

kennytm committed:
Rollup merge of #49381 - withoutboats:str_unicode, r=SimonSapin

Add is_whitespace and is_alphanumeric to str.

The other methods from `UnicodeStr` are already stable inherent
methods on str, but these have not been included.

r? @SimonSapin

WiSaGaN committed:
Add back 1.24.1 release notes

kennytm committed:
Rollup merge of #49395 - petrochenkov:obsolete, r=alexcrichton

libsyntax: Remove obsolete.rs

This little piece of infra is obsolete (ha-ha) and is unlikely to be used in the future, even if new obsolete syntax appears.

kennytm committed:
Rollup merge of #49399 - ehuss:termcolor-update, r=alexcrichton

Fix diagnostic colors on Windows 10 console.

This updates termcolor to pick up BurntSushi/ripgrep#867.

Fixes #49322.

kennytm committed:
Rollup merge of #49401 - alercah:format, r=cramertj

Add missing '?' to format grammar.

rust-highfive commented:
r? @dtolnay

(rust_highfive has picked a reviewer for you, use r? to override)

kennytm commented:
@bors r+ p=8

bors commented:
:pushpin: Commit b4bc2b0 has been approved by `kennytm`

<!-- @bors r=kennytm b4bc2b0cf8a889bf0c1e4fdd826122ec5b4a6479 -->

TimNN committed:
Update compiler-rt with fix for 32bit iOS ARM

kennytm committed:
Rollup merge of #49417 - TimNN:fix-ios, r=alexcrichton

Update compiler-rt with fix for 32bit iOS ARM

kennytm commented:
@bors r+

Added #49417.

bors commented:
:pushpin: Commit 0873974 has been approved by `kennytm`

<!-- @bors r=kennytm 0873974290f4e68b9e12c6922f07c0a7c8a32c98 -->

lukaslueg committed:
Update CONTRIBUTING.md

The current link is a 404, just link to the main repo page

kennytm committed:
Rollup merge of #49407 - WiSaGaN:patch-1, r=alexcrichton

Add back 1.24.1 release notes

kennytm committed:
Rollup merge of #49426 - lukaslueg:patch-1, r=kennytm

Update CONTRIBUTING.md

The current link is a 404, just link to the main repo page

kennytm commented:
@bors r+

Added #49202, #49426

bors commented:
:pushpin: Commit 605ea7c has been approved by `kennytm`

<!-- @bors r=kennytm 605ea7c31f7341995c2d1ae12b4b33fe6bd908b5 -->

bors commented:
:hourglass: Testing commit 605ea7c31f7341995c2d1ae12b4b33fe6bd908b5 with merge 35f368f0823abe38db060a3a7ba764da3410ee63...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/359074134?utm_source=github_status&utm_medium=notification)

kennytm commented:
@bors retry 

A macOS job is stuck at downloading the `sccache` binary.

bors commented:
:hourglass: Testing commit 605ea7c31f7341995c2d1ae12b4b33fe6bd908b5 with merge 59ec9bfc66b86f04b50f00bb32839315f59252ec...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.6833), [status-travis](https://travis-ci.org/rust-lang/rust/builds/359186078?utm_source=github_status&utm_medium=notification)
Approved by: kennytm
Pushing 59ec9bfc66b86f04b50f00bb32839315f59252ec to master...
